### PR TITLE
Custom import sources

### LINKED ISSUE #603: Customization of base_paths for relative imports
Currently, the `base_paths` for relative grammar imports only base on following, if I understood correctly:

1. the directory where the grammar file(-like object) is (fetching from the grammar file path)
2. the directory where the main script is (fetching from `sys.modules['__main__'].__file__`

However, this limits the architecture of projects that are using `lark`.

Imagine we have the following architecture:

```
|- package/
    |- submodule1/
        |- parser.py            # with a string grammar wanting to import some 
                                # shared rules/terminals from the shared grammar
    |- submodule2/              # maybe I have another submodule that wants to 
                                # import some rules/terminals from the shared grammar
    |- shared/
        |- grammar/
            |- shared.lark      # some shared rules/terminals
                                # say we have a terminal `INT` defined for the example below
```

Well, we could probably use 2) listed above to do the relative imports?

```lark
// not sure if this works
%import .shared.grammar.shared.INT 
```

However, this doesn't run if the package is not running independently. For example, using `pytest`, since `sys.modules['__main__'].__file__` will become the path to `pytest`.

**Ideas/Proposals:**

1. Reading paths from an env variable, say `LARK_GRAMMAR_PATH` (just like `PYTHON_PATH`)
2. Opening the `source` argument from `Lark` construct to allow overridden from outside.
3. Allowing pre-loaded grammar by `load_grammar.load_grammar`, since this function already allows specification of `grammar_name`, where the `base_paths` can fetch from

    With the above architecture, say in `parser.py`, we can do:
    ```python
    from pathlib import Path
    SHARED_GRAMMER= str(Path(__file__).parent.parent.joinpath('shared', 'grammar', 'shared.lark').resolve())
    grammar = load_grammar.load_grammar(
        r"""
        start: INT

        %import .shared.INT
        """,
        SHARED_GRAMMER
    )
    parser = Lark(grammar, ...)
    ```
    
    Then in `Lark` construct, we need to tell if the grammar is pre-loaded:
    ```python
    from load_grammer import Grammar

    class Lark:
        def __init__(self, grammar, ...):
             if isinstance(grammar, Grammar):
                 self.grammar = grammar
                 self.source = '<pre-loaded>'
                 # may have to deal with caching separately in such case
             else:
                 # regular grammar loading
    ```

### PR DESCRIPTION
**Pull request by MegaIng**
Added a new option to `Lark`, `import_sources`.
This is a list of either:

- A string, representing a folder which will be directly searched for any grammars
- A callable of the form `(base_paths, grammar_path) -> (grammar_name, grammar_source)`
  - `base_paths`: A list of paths lark identified as source folders. Probably not useful in custom loaders
  - `grammar_path`: The joined relative path of the grammar file, including the extension
  - `grammar_name`: The name that will be used for the grammar. Should be the path, but doesn't have to.
  - `grammar_source`: The raw grammar source, not a file object.

(This branch also includes a change to use `pkgutil` for `common.lark`, closing #703)

This should also close #603.

Note: while not a draft, I would like for someone to check whether this works for zipapp before merging.

### CODE REVIEW

**Comment by MegaIng** [comment]
@sinoroc

This is how you could write a zipfile aware loader:
https://github.com/lark-parser/lark/pull/704/files#diff-ed9d66898a1188fc8f00f74cc941be36R1786

Does this help for your case?

**Comment by erezsh** [comment]
You can rewrite `import_grammar` without the code duplication, by putting the `if callable()` inside the `with`.

**Comment by sinoroc** [comment]
Thanks! I didn't expect anything so soon :) 

I will give it a try. I'm new to _Poetry_'s code base, so it will take me a while. Plus there is the additional difficulty of _lark_ being vendored in, so it will take me a bit to figure out how to test this. I will let you know...

**Comment by MegaIng** [comment]
Ok, after thinking a bit more about this, I decided that this should actually still be a draft:

- I am pretty sure relative import currently don't work
- I would like to add a simple interface so that it is easy to have a custom loader that uses grammar from a specific package.

**Comment by erezsh** [comment]
Don't we have tests for relative imports? Or do you mean it doesn't work for callable sources?

**Comment by MegaIng** [comment]
Yes, for the normal imports it works. But I am pretty sure it doesn't quite work for the current 'model' implementation. (And yes, we have tests. But I honestly don't have an overview on how they are organized. Maybe you can write a few tests that test all possible edge-cases with relative imports for custom sources? I don't think we need tests for renaming though)

**Comment by MegaIng** [comment]
Ok, with this commit,  I think everything works. I also tested it from inside a zip files and everything works (Both `common.lark` loading and custom package imports) Maybe a few more docs would be nice, but where?

**Comment by MegaIng** [comment]
@sinoroc Also note that for your usecase it should now be as simple as calling `Lark.open_from_package(__name__, 'grammar_name.lark', ('grammar_folder', ))`

File: lark/load_grammar.py [hunk 1]
```diff
@@ -648,17 +649,67 @@ def compile(self, start):
...
+        self.search_paths = search_paths
+        
+    def __repr__(self):
+        return "%s(%r, %r)" % (type(self).__name__, self.pkg_name, self.search_paths)
+
+    def __call__(self, base_paths, grammar_path):
+        if len(base_paths) == 0:
+            to_try = self.search_paths
+        else:
+            assert len(base_paths) == 1
```
Comment (erezsh) [inline lark/load_grammar.py:673]: Why is `len(base_paths)` only 0 or 1? Maybe it should be `Optional[str]` instead then?

File: lark/load_grammar.py [hunk 2]
```diff
@@ -648,17 +649,67 @@ def compile(self, start):
...
+            if not base_paths[0].startswith('<%s:' % (self.pkg_name,)):
```
Comment (erezsh) [inline lark/load_grammar.py:674]: What is this?

File: lark/load_grammar.py [hunk 3]
```diff
@@ -648,17 +649,67 @@ def compile(self, start):
...
-                with open(joined_path, encoding='utf8') as f:
-                    text = f.read()
-                grammar = load_grammar(text, joined_path, re_)
+                if callable(source):
+                    joined_path, text = source(base_paths, grammar_path)
+                else:
+                    joined_path = os.path.join(source, grammar_path)
+                    with open(joined_path, encoding='utf8') as f:
+                        text = f.read()
+            if text is not None:
```
Comment (erezsh) [inline lark/load_grammar.py:709]: Why would `text` be None?

File: lark/lark.py [hunk 4]
```diff
@@ -401,6 +410,28 @@ def open(cls, grammar_filename, rel_to=None, **options):
             grammar_filename = os.path.join(basepath, grammar_filename)
         with open(grammar_filename, encoding='utf8') as f:
             return cls(f, **options)
+    
+    @classmethod
+    def open_from_package(cls, package, grammar_path, search_paths=("",), **options):
+        """Create an instance of Lark with the grammar loaded from within the package `package`.
+        This allows grammar loading from zipapps.
+        
+        Will also create a `FromPackageLoader` instance and add it to the `import_sources` to simplify importing
```
Comment (erezsh) [inline lark/lark.py:419]: Why is this useful for the user to know?

File: lark/lark.py [hunk 5]
```diff
@@ -115,6 +119,8 @@ class LarkOptions(Serialize):
         'edit_terminals': None,
         'g_regex_flags': 0,
         'use_bytes': False,
+        'import_sources': [],
```
Comment (erezsh) [inline lark/lark.py:122]: The common name for this is `path`, both in windows, linux, and Python with `sys.path`.

So maybe `import_paths` makes more sense?

File: lark/lark.py [hunk 6]
```diff
@@ -115,6 +119,8 @@ class LarkOptions(Serialize):
...
+        'source': None,
```
Comment (erezsh) [inline lark/lark.py:123]: `source` is a little vague in this context. Maybe call it `source_name` or `source_path`.

**Review by erezsh — CHANGES_REQUESTED** [review]
Left you some comments.

File: lark/lark.py (hunk 4 above)
Comment (MegaIng) [reply lark/lark.py:419]: I thought that it would be useful to prevent users from manually specific the `import_sources`. `FromPackageLoader` is after all a public class.

File: lark/lark.py (hunk 6 above)
Comment (MegaIng) [reply lark/lark.py:123]: Yes, true

File: lark/lark.py (hunk 5 above)
Comment (MegaIng) [reply lark/lark.py:122]: I though so too. But those paths are almost always just file paths, not custom loaders. If you want, we can rename it, but I specifically decided against it.

File: lark/load_grammar.py (hunk 2 above)
Comment (MegaIng) [reply lark/load_grammar.py:674]: This is to recognize relative imports from within a package. When a grammar is loaded by `FromPackageLoader`, it gets a path in the form `<{pkg_name}:/{full_path}>` Where `full_path` is the argument that was passed to `pkgutil.get_data`. Here we unparse it to get the relative path that was specified in importing (allowing relative imports). I will write a comment for this.

File: lark/load_grammar.py (hunk 1 above)
Comment (MegaIng) [reply lark/load_grammar.py:673]: This is the design you choose in the implementation. I have nothing against changing this, but I imagine you at some point in the future might want to add additional elements to `base_paths`. If that is the case, this `assert` will trigger and we have to change something here. I could future proof this now if you want. (e.g. allow multiple `base_paths` entries.

File: lark/lark.py (hunk 4 above)
Comment (erezsh) [reply lark/lark.py:419]: Hmm maybe it's better to say something like 

    Imports in the grammar will use the `package` and `search_paths` provided, through `FromPackageLoader`.

But I'll leave it to your judgement.

File: lark/lark.py (hunk 5 above)
Comment (erezsh) [reply lark/lark.py:122]: Even so, `import_paths` seems more adequate, no?

File: lark/load_grammar.py (hunk 1 above)
Comment (erezsh) [reply lark/load_grammar.py:673]: When I used `base_paths`, I did not differentiate between base_paths and search_paths, and I did not consider relative imports vs absolute imports.

Now that we make the distinction, there shouldn't be more than one base_path (or so it seems to me). And search_paths has now taken its place for a list of paths. So I think it should be turned into an optional string.

File: lark/lark.py (hunk 6 above)
Comment (MegaIng) [reply lark/lark.py:123]: Would you be ok with also renaming `Lark.source`? It is as vague. (Also I would rename `Lark.grammar_source` to `Lark.source_code`)

File: lark/load_grammar.py (hunk 2 above)
Comment (erezsh) [reply lark/load_grammar.py:674]: Why are we accepting this repr-like string? Just so we can support `__name__`?

File: lark/lark.py (hunk 6 above)
Comment (erezsh) [reply lark/lark.py:123]: In principle I agree, but it would break compatibility. You can open an issue for this for v1.0

Fwiw, at least when you read the existing `source` property, you can understand what it is from its value.

File: lark/load_grammar.py (hunk 2 above)
Comment (MegaIng) [reply lark/load_grammar.py:674]: No, because there might not be a meaningful way of expressing the location of the grammar file as a normal file path the `open` accepts. This is similar to `<string>`, just that it also carries information for imports inside of it instead of implied/inferred at a latter stage. This is a format I designed to just be used internally by the `FromPackageLoader` and also be readable by the end user. Do you have an alternative suggestion?

File: lark/lark.py (hunk 6 above)
Comment (MegaIng) [reply lark/lark.py:123]: I would of course add backwards compatibility via properties.

File: lark/load_grammar.py (hunk 2 above)
Comment (erezsh) [reply lark/load_grammar.py:674]: Well, if you already know the package name.. Why not just accept a regular path within the package?

Comment (MegaIng) [reply lark/load_grammar.py:674]: Because that could cause problems when there are multiple packages from which we are loading. And it is not like these paths would actually be useful for anything else.

File: lark/load_grammar.py (hunk 3 above)
Comment (MegaIng) [reply lark/load_grammar.py:709]: Because the `with suppress()` statement can fail and then the previous value of text will stay. I assign `None` to it before hand. Alternatively, I can rework it into a `try except else` statement if you want

File: lark/lark.py (hunk 6 above)
Comment (erezsh) [reply lark/lark.py:123]: Okay, I suppose there's no harm

File: lark/load_grammar.py (hunk 2 above)
Comment (erezsh) [reply lark/load_grammar.py:674]: Maybe I missing something, but this all seems a little messier than it has to be.

But if you insist on this hack, parse it with a regexp. It shouldn't take more than two lines.

And document it well.

File: lark/load_grammar.py (hunk 3 above)
Comment (erezsh) [reply lark/load_grammar.py:709]: Yes, that's why I loaded the text inside the suppress. But if you think it's better outside (I understand, imports and such), it's better to use `try.. except`

Comment (MegaIng) [reply lark/load_grammar.py:709]: Read the next comment.

Comment (erezsh) [reply lark/load_grammar.py:709]: I already said I understand. What did that comment add to *my* comment?

File: lark/load_grammar.py (hunk 2 above)
Comment (MegaIng) [reply lark/load_grammar.py:674]: If you have an alternative suggestion that doesn't run into problems when there are multiple files with the same name in different places, I am open to ideas.

File: lark/load_grammar.py (hunk 3 above)
Comment (MegaIng) [reply lark/load_grammar.py:709]: Sorry, I misread your comment.

File: lark/load_grammar.py (hunk 2 above)
Comment (erezsh) [reply lark/load_grammar.py:674]: So, I'm trying to think of the actual use cases. Help me understand.

The first use case, is when trying to load the main grammar from a package. But there is a method for it, and it's only one grammar, one package. No difficulty.

The second use case, is when trying to import a grammar from within a grammar. In that case, we currently don't allow to specify the package anyway. We only allow a dotted path.

I can't think of other use cases.

So, in what realistic situation will this feature be used?

File: lark/lark.py (hunk 5 above)
Comment (MegaIng) [reply lark/lark.py:122]: Python uses `sys.path` and `sys.path_hooks` for this. I don't want to split it into multiple arguments.

Ok, I will rename it `import_paths`.

File: lark/load_grammar.py (hunk 2 above)
Comment (MegaIng) [reply lark/load_grammar.py:674]: Maybe one can create a package with grammar library with lots of different grammar files. Then you can add that package to your own module where you also have a somewhat complex grammar split over multiple files. (Or a plugin system). If you try to use relative imports, we have to prevent the grammars from one package to find the grammars in the other package.

Under the old file-based system, this wasn't a problem since we always specified an absolute file path that clearly defines which grammar file we are talking about.

But now we can't easily do that. If we just use `full_path` there might be ambiguity of both packages use the `grammars` folder. If we try to use an absolute path, this will break when we are inside of a zip file. So we somehow have to encode both the package and the path inside of the package, unless we don't want to support those usecases. And I don't see any real harm done by this.

Comment (erezsh) [reply lark/load_grammar.py]: You are describing a situation that no one currently has.

Also, I still don't understand who will pass the `<package:path>` string, and where it will pass.

If both packages have a `grammars` folder in the search paths, that is still not a collision. Only if they have the same `whatever.lark` file it will collide, no? And then it will choose whatever is first on the list.

The harm is code bloat. Death by a million cuts. Every little feature we add, is something we will have to support. And one line after another end up being hundreds of lines, that force us to do things in a certain way, because we overengineered when we didn't have to.

Again, maybe if you can give a specific use case that it will solve, it will help me understand. If you can't explain it, show me the code from the user's perspective.

Comment (MegaIng) [reply lark/load_grammar.py]: > Also, I still don't understand who will pass the <package:path> string, and where it will pass.
This is the end of the function. I added a few more comments. So it is just the `FromPackageLoader` talking to itself.

Comment (erezsh) [reply lark/load_grammar.py]: I see. If this is an internal structure, and not something that is user-facing, then please let's use a dedicated class for it.

Comment (MegaIng) [reply lark/load_grammar.py]: Well, it is also the name of the grammar to be used in error reporting (the source argument to `GrammarLoader.load_grammar`) The benefit of just using str is that we don't need to make any changes in `load_grammar`.

Comment (erezsh) [reply lark/load_grammar.py]: That's what `__str__` is for. And if we need to add a `str()` call somewhere, it will be well worth it.

Comment (MegaIng) [reply lark/load_grammar.py]: Well, I pushed another commit. I would call this more code bloat than before, but if you prefer this, ok.

File: lark-stubs/lark.pyi [hunk 7]
```diff
@@ -32,14 +32,20 @@ class LarkOptions:
...
+    pkg_name: str
+    path: str
+    
+    def __init__(self, pkg_name: str, path: str):
+
 class FromPackageLoader:
     def __init__(self, pkg_name: str, search_paths: Tuple[str, ...] = ...): ...
     
-    def __call__(self, base_paths: str, grammar_path: str) -> Tuple[str, str]: ...
+    def __call__(self, base_paths: Union[None, str, PackageResource], grammar_path: str) -> Tuple[PackageResource, str]: ...
```
Comment (erezsh) [inline lark-stubs/lark.pyi:48]: This should be `base_path`

File: lark-stubs/lark.pyi [hunk 8]
```diff
@@ -32,11 +32,25 @@ class LarkOptions:
...
+    def __init__(self, pkg_name: str, search_paths: Tuple[str, ...] = ...): ...
+    
+    def __call__(self, base_path: Union[None, str, PackageResource], grammar_path: str) -> Tuple[PackageResource, str]: ...
 
 
 class Lark:
-    source: str
-    grammar_source: str
+    source_path: str
+    source_code: str
```
Comment (erezsh) [inline lark-stubs/lark.pyi:53]: Sorry for not catching this earlier, but this should be `source_text`. Not everyone uses lark just to parse code.

File: lark/lark.py [hunk 9]
```diff
@@ -450,6 +479,15 @@ def parse(self, text, start=None, on_error=None):
                         # Prevent infinite loop
                         raise e2
                     e = e2
+    
+    @property
+    def source(self):
+        warn("Lark.source attribute has been renamed to Lark.source_path", DeprecationWarning)
+        return self.source_path
+    
+    @source.setter
```
Comment (erezsh) [inline lark/lark.py:488]: This is read-only. A setter isn't necessary.

File: lark-stubs/lark.pyi (hunk 8 above)
Comment (MegaIng) [reply lark-stubs/lark.pyi:53]: Well, it is the source code for the grammar. Maybe `source_grammar`?

Comment (erezsh) [reply lark-stubs/lark.pyi:53]: Hmm then my comment is irrelevant. But yes, maybe that's a better name anyway.

**Review by erezsh — APPROVED** [review]

**Comment by erezsh** [comment]
`grammar_source` is removed, but needs property+warning for backward compatibility.

I think that's the last hold out.

**Comment by erezsh** [comment]
Thanks for the PR!

File: lark/load_grammar.py (hunk 2 above)
Comment (erezsh) [reply lark/load_grammar.py]: To be fair, this whole zipapp feature was a code bloat :)
